homeassistant/components/oralb/__init__.py

```diff
@@ -1,30 +1,104 @@
 """The OralB integration."""
 from __future__ import annotations
 
+import asyncio
+from datetime import timedelta
+import logging
+
+import async_timeout
+from oralb import OralB
+
+from homeassistant.components import bluetooth
+from homeassistant.components.bluetooth.match import ADDRESS, BluetoothCallbackMatcher
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import Platform
-from homeassistant.core import HomeAssistant
+from homeassistant.const import CONF_MAC, EVENT_HOMEASSISTANT_STOP, Platform
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
-from .const import DOMAIN
+from .const import DEVICE_TIMEOUT, DOMAIN, UPDATE_SECONDS
+from .models import OralBData
 
 # For your initial PR, limit it to 1 platform.
 PLATFORMS: list[Platform] = [Platform.SENSOR]
+_LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up OralB from a config entry."""
+    address: str = entry.data[CONF_MAC]
+    print("using address: ", address)
+    ble_device = bluetooth.async_ble_device_from_address(hass, address.upper(), True)
+    if not ble_device:
+        raise ConfigEntryNotReady(f"Could not find OralB device with address {address}")
+
+    orlb = OralB(ble_device)
+
+    @callback
+    def _async_update_ble(
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a ble callback."""
+        orlb.set_ble_device(service_info.device)
+
+    entry.async_on_unload(
+        bluetooth.async_register_callback(
+            hass,
+            _async_update_ble,
+            BluetoothCallbackMatcher({ADDRESS: address}),
+            bluetooth.BluetoothScanningMode.PASSIVE,
+        )
+    )
+
+    async def _async_update():
+        """Update the device state."""
+        await orlb.gatherdata()
```

Comment:
you need to catch Bluetooth specific exceptions and raise `UpdateFailed` exceptions instead.